File: src/pdf.ts
import { mkdtempSync, readFileSync, readdirSync, rmSync, writeFileSync } from "node:fs";
import { tmpdir } from "node:os";
import { join } from "node:path";
import { spawnSync } from "node:child_process";

export interface PdfRunResult { status: number | null; stdout: Buffer; stderr: Buffer; error?: Error }
export type PdfRunner = (command: string, args: string[], input?: Buffer) => PdfRunResult;
export interface PdfExtraction { text: string; pageCount: number | null; ocrUsed: boolean; extractor: "pdftotext" | "tesseract" }

const defaultRunner: PdfRunner = (command, args, input) => {
  const result = spawnSync(command, args, { input, timeout: 60_000, maxBuffer: 32 * 1024 * 1024 });
  return { status: result.status, stdout: Buffer.from(result.stdout ?? []), stderr: Buffer.from(result.stderr ?? []), ...(result.error ? { error: result.error } : {}) };
};

function execute(runner: PdfRunner, command: string, args: string[], input?: Buffer): Buffer {
  const result = runner(command, args, input);
  if (result.error || result.status !== 0) throw new Error(`PDF tool failed: ${command}: ${result.error?.message ?? result.stderr.toString("utf8").slice(0, 500)}`);
  return result.stdout;
}

export function extractPdf(content: Buffer, options: { runner?: PdfRunner; maxBytes?: number; maxPages?: number; minTextChars?: number; ocrLanguage?: string } = {}): PdfExtraction {
  if (!content.subarray(0, 5).equals(Buffer.from("%PDF-"))) throw new Error("Content is not a valid PDF");
  if (content.length > (options.maxBytes ?? 20_000_000)) throw new Error("PDF exceeds size limit");
  const runner = options.runner ?? defaultRunner;
  const direct = execute(runner, "pdftotext", ["-layout", "-", "-"], content).toString("utf8").trim();
  if (direct.length >= (options.minTextChars ?? 200)) return { text: direct, pageCount: null, ocrUsed: false, extractor: "pdftotext" };
  const directory = mkdtempSync(join(tmpdir(), "egypt-research-pdf-"));
  try {
    const pdfPath = join(directory, "input.pdf");
    writeFileSync(pdfPath, content);
    const info = execute(runner, "pdfinfo", [pdfPath]).toString("utf8");
    const pageCount = Number(/^Pages:\s+(\d+)/m.exec(info)?.[1] ?? 0);
    if (pageCount < 1 || pageCount > (options.maxPages ?? 200)) throw new Error("PDF page count is outside allowed range");
    const prefix = join(directory, "page");
    execute(runner, "pdftoppm", ["-png", "-r", "200", "-f", "1", "-l", String(pageCount), pdfPath, prefix]);
    const pages = readdirSync(directory).filter((name) => /^page-\d+\.png$/.test(name)).sort((left, right) => Number(left.match(/\d+/)?.[0]) - Number(right.match(/\d+/)?.[0]));
    const text = pages.map((page) => execute(runner, "tesseract", [join(directory, page), "stdout", "-l", options.ocrLanguage ?? "ara+eng", "--psm", "6"], readFileSync(join(directory, page))).toString("utf8").trim()).filter(Boolean).join("\n\n");
    return { text, pageCount, ocrUsed: true, extractor: "tesseract" };
  } finally {
    rmSync(directory, { recursive: true, force: true });
  }
}

File: src/egypt_research_mcp/pdf.py
from __future__ import annotations

import re
import subprocess
import tempfile
from dataclasses import dataclass
from pathlib import Path
from typing import Callable


Runner = Callable[..., subprocess.CompletedProcess]


@dataclass(frozen=True)
class PdfExtraction:
    text: str
    page_count: int | None
    ocr_used: bool
    extractor: str


def _execute(runner: Runner, args: list[str], **kwargs) -> subprocess.CompletedProcess:
    try:
        result = runner(args, capture_output=True, timeout=60, **kwargs)
    except (FileNotFoundError, subprocess.TimeoutExpired) as error:
        raise RuntimeError(f"PDF tool failed: {args[0]}") from error
    if result.returncode != 0:
        message = result.stderr.decode(errors="replace")[:500]
        raise RuntimeError(f"PDF tool failed: {args[0]}: {message}")
    return result


def extract_pdf(
    content: bytes,
    *,
    runner: Runner = subprocess.run,
    max_bytes: int = 20_000_000,
    max_pages: int = 200,
    min_text_chars: int = 200,
    ocr_language: str = "ara+eng",
) -> PdfExtraction:
    if not content.startswith(b"%PDF-"):
        raise ValueError("Content is not a valid PDF")
    if len(content) > max_bytes:
        raise ValueError("PDF exceeds size limit")

    direct = _execute(
        runner, ["pdftotext", "-layout", "-", "-"], input=content
    ).stdout.decode(errors="replace").strip()
    if len(direct) >= min_text_chars:
        return PdfExtraction(direct, None, False, "pdftotext")

    with tempfile.TemporaryDirectory(prefix="egypt-research-pdf-") as directory:
        pdf_path = Path(directory) / "input.pdf"
        pdf_path.write_bytes(content)
        info = _execute(runner, ["pdfinfo", str(pdf_path)]).stdout.decode(
            errors="replace"
        )
        match = re.search(r"^Pages:\s+(\d+)", info, flags=re.MULTILINE)
        page_count = int(match.group(1)) if match else 0
        if page_count < 1 or page_count > max_pages:
            raise ValueError("PDF page count is outside allowed range")
        prefix = Path(directory) / "page"
        _execute(
            runner,
            [
                "pdftoppm",
                "-png",
                "-r",
                "200",
                "-f",
                "1",
                "-l",
                str(page_count),
                str(pdf_path),
                str(prefix),
            ],
        )
        pages = []
        for image in sorted(Path(directory).glob("page-*.png")):
            result = _execute(
                runner,
                ["tesseract", str(image), "stdout", "-l", ocr_language, "--psm", "6"],
            )
            pages.append(result.stdout.decode(errors="replace").strip())
        return PdfExtraction("\n\n".join(filter(None, pages)), page_count, True, "tesseract")
